Backend MVP

tanyginvv opened the pull request:
Расписал команды и запросы к существующим сущностям

tanyginvv committed:
начал настройку EF

создал таблицу Recipes, расписал Domain

tanyginvv committed:
сконфигурировал бд с помощью EntityFramework

tanyginvv committed:
поменял название Recipes.Application и удалил ненужную папку

tanyginvv committed:
правки в папках

tanyginvv committed:
Реализовал CQRS(для рецепта), настраиваю апи

IServiceCollection не подключается, апи запускается с ошибками( или не запускается)

tanyginvv committed:
Расписал CQRS для всех сущностей

tanyginvv committed:
дописал bindings у сущностей

tanyginvv committed:
Настроил  базу CQRS для Api, начинаю дописывать полностью

tanyginvv committed:
Настроил получение всех рецептов

tanyginvv committed:
реализовал добавление рецепта

tanyginvv committed:
реализовал удаление рецепта

удаляются сам рецепт, его шаги, ингредиенты, теги остаются нетронуты, удаляется только связь тег-рецепт

tanyginvv committed:
правки по queries

tanyginvv committed:
Реализовал изменение рецепта

tanyginvv committed:
Убрал лишние комментарии

tanyginvv committed:
убрал лишние миграции, оставил одну

tanyginvv committed:
Вынес интерфейсы репозитории в инфраструктуру, потому что образовался вечный цикл в зависимостях, соответственно правки в пространствах имен

tanyginvv committed:
Вынес зависимости репозиториев в Bindings

tanyginvv committed:
Раскидал биндинги по проектам

tanyginvv committed:
убрал большое количество миграций, оставил одну

tanyginvv committed:
вынес интерфейсы репозиториев в Application, вернул enable у nullable

tanyginvv committed:
вынес интерфейсы репозиториев в Application

tanyginvv committed:
правки по пространствам имен

tanyginvv committed:
вынес DI-бд к биндинг инфраструктуры

tanyginvv committed:
убрал из ветки фронтенд

VladDotMe commented:
Папку Recipes переименовать -> Recipes.WebAPI (и проект тоже лучше в WebAPI для большей конкретики)

VladDotMe commented on backend/Recipes/Recipes/Program.cs:
```diff
@@ -1,25 +1,25 @@
-// Add services to the container.
+var builder = WebApplication.CreateBuilder( args );
 
+// ����������� ������ ������������
```
неправильная кодировка файла

tanyginvv replied:
уберу этот комментарий, а так гитхаб не поддерживает вроде русские буквы, поэтому так

VladDotMe replied:
> уберу этот комментарий, а так гитхаб не поддерживает вроде русские буквы, поэтому так

Поддерживает, только файл должен быть в UTF-8

VladDotMe commented on backend/Recipes/Recipes/Program.cs:
```diff
@@ -1,25 +1,25 @@
+var builder = WebApplication.CreateBuilder( args );
 
+// ����������� ������ ������������
 builder.Services.AddControllers();
```
давай какой-нибудь порядок что-ли, AddControllers после Application/Infrastructure Bindings

VladDotMe commented on backend/Recipes/Recipes.Domain/Entities/Ingredient.cs:
```diff
@@ -22,5 +24,11 @@ public void SetDescription( string description )
             Description = description;
         }
+
+        public void SetRecipe( Recipe recipe )
```
Не согласен с этими методами, давай все Set методы уберем во всех Entities. Если тебе нужно устанавливать значение пропертей - так сделай им публичные сеттеры.

VladDotMe commented on backend/Recipes/Recipes.Application/CQRSInterfaces/ICommandHandler.cs:
```diff
@@ -0,0 +1,17 @@
+        Task<CommandResult> HandleAsync( TCommand command );
+    }
+
+    public interface ICommandHandler<TResult, TCommand>
```
не используется

VladDotMe commented on backend/Recipes/Recipes.Application/CQRSInterfaces/IQueriesHandler.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using Recipes.Application.Recipes.Dtos;
```
Почему файл называется IQueriesHandler а не IQueryHandler?

VladDotMe commented on backend/Recipes/Recipes.Application/Repositories/BaseRepositories/IAddedRepositories.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Recipes.Application.Repositories.BaseRepositories
```
Не используется, и все репозитории в папке BaseRepositories тоже. Впринципе можно удалить, не вижу для них юзкейсов..

VladDotMe commented on backend/Recipes/Recipes.Application/Recipes/Commands/CreateRecipe/CreateRecipeCommand.cs:
```diff
@@ -0,0 +1,19 @@
+﻿using Recipes.Application.Ingredients.Dtos;
```
Мб все эти папки с логикой (Recipes, Steps, Tags, Ingredients) запихнуть в общую папку UseCases? Что думаешь?

tanyginvv replied:
думал об этом, пожалуй так и сделаю

VladDotMe commented on backend/Recipes/Recipes.Application/Steps/Queries/GetStepsByRecipeIdQuery/GetStepsByRecipeIdQueryHandler.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace Recipes.Application.Steps.Queries.GetStepsByRecipeIdQuery
+{
+    public class GetStepsByRecipeIdQueryHandler : IQueryHandler<GetStepsByRecipeIdQueryDto, GetStepsByRecipeIdQuery>
```
А го использовать везде primary constructors? На .NET8 все таки пишем..) Не заставляю, в целом субъективщина, но мне кажется, что повышается читаемость кода. 

UPD: Решили, что используем primary constructors везде.

VladDotMe commented on backend/Recipes/Recipes/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,105 @@
+        private readonly IQueryHandler<GetRecipeByIdQueryDto, GetRecipeByIdQuery> _getRecipeByIdQueryHandler;
+        private readonly IQueryHandler<IEnumerable<RecipeDto>, GetAllRecipesQuery> _getAllRecipesQueryHandler;
+
+        public RecipesController(
```
го тож primary constructor)

VladDotMe commented on backend/Recipes/Recipes.Infrasructure/InfrastructureBinding.cs:
убери это из диффа, если не знаешь как - пиши, разберемся

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Recipes.Infrastructure.csproj:
```diff
@@ -0,0 +1,31 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Recipes.Application\Recipes.Applicatio.csproj" />
```
тут лишняя хрень, не компилится

VladDotMe commented on backend/Recipes/Recipes/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> CreateRecipe( [FromBody] CreateRecipeCommand command )
```
Что-то здесь не так.. Почему в методах Api используются сущности Application? Так нельзя, здесь должны быть Dto. Далее эти DTO надо перегонять в сущности Application.  

Кстати, если не хочешь вручную конвертить сущности туда сюда - то для автоматической перегонки можно использовать Automapper/Mapster или другую аналогичную библиотеку. У нас в компании не особо практикуется, но вообще подход хороший.
 
В целом оставлю на твое усмотрение: использовать автомапперы или перегонять сущности вручную.

VladDotMe replied:
Еще хорошей практикой является валидация сущностей, которые подаются на вход на бекенд. То есть надо проверить, что там есть все необходимые поля и они имеют нужные значения. 
Можно использовать дефолтные шарповые https://learn.microsoft.com/en-us/aspnet/mvc/overview/older-versions-1/models-data/validation-with-the-data-annotation-validators-cs

либо библиотеку FluentValidation
https://github.com/FluentValidation/FluentValidation

tanyginvv replied:
Попробовал сделать с помощью AutoMapper, вылезла ошибка, что нет настроенных mapperов и пока решил оставить вручную

VladDotMe commented on backend/Recipes/Recipes/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,105 @@
+            {
+                return BadRequest( result.ValidationResult.Error );
+            }
+            return NoContent();
```
Перед return обычно оставляем пустую строку

VladDotMe commented on backend/Recipes/Recipes/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,105 @@
+            {
+                return BadRequest( result.ValidationResult.Error );
+            }
+            return NoContent();
```
Здесь лучше вернуть CreatedAtAction и айдишник созданного рецепта.

tanyginvv replied:
Проблематично возвращать id, тк это команда, а не query. Команды могут возвращать только Result

VladDotMe replied:
> Проблематично возвращать id, тк это команда, а не query. Команды могут возвращать только Result

Так сделай, чтобы команда тоже возвращала generic Result

VladDotMe commented on backend/Recipes/Recipes/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        [HttpDelete( "{id}" )]
+        public async Task<IActionResult> DeleteRecipe( [FromRoute] int id )
```
здесь и в других местах тоже лучше валидировать подаваемые на вход значения через атрибуты

VladDotMe commented on backend/Recipes/Recipes/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,105 @@
+        [HttpGet( "{id}" )]
+        public async Task<IActionResult> GetRecipeById( [FromRoute] int id )
+        {
+            var result = await _getRecipeByIdQueryHandler.HandleAsync( new GetRecipeByIdQuery { Id = id } );
```
А еще в новом дотнете ты вроде можешь не в конструкторе а непосредственно в нужном контроллере прокидывать нужный тебе хендлер.  Так лучше, чтобы они были доступны только там, где они используются, а не во всем классе. 

        public async Task<IActionResult> GetRecipeById( [FromRoute] int id,
            [FromServices]IQueryHandler<GetRecipeByIdQueryDto, GetRecipeByIdQuery> getRecipeByIdQueryHandler )

VladDotMe commented on backend/Recipes/Recipes.Application/Results/QueryResult.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace Recipes.Application.Results
+{
+    public class QueryResult<TQueryResultData> where TQueryResultData : class
```
Чем принципиально отличается QueryResult от CommandResult? 

У тебя здесь какая-то разновидность паттерна Result но сильно усложненная, зачем?

Можешь почитать про паттерн тут:
https://medium.com/@wgyxxbf/result-pattern-a01729f42f8c

Предлагаю убрать QueryResult, CommandResult, ValidationResult и заюзать один класс Result по канонам из учебников. 

Для примера - ниже код Result (взял из одной либы и чутка доработал для наших реалий), можешь брать целиком, если что-то будет не нужно - удали или доработай, но в целом это хороший код. 
и пример использования.
```
namespace Recipes.Application.Results
{
    public record ResultError( string Message );

    public class Result
    {
        public readonly ResultError? Error;

        public bool IsSuccess => Error is null;

        public static Result Success { get; } = FromSuccess();

        private Result( ResultError? error )
        {
            Error = error;
        }

        public static Result FromSuccess()
        {
            return new Result( null );
        }

        public static Result FromError( ResultError? error )
        {
            return new Result( error );
        }

        public static Result FromError( string errorText )
        {
            return new Result( new ResultError( errorText ) );
        }
    }

    public class Result<T>
    {
        public readonly T? Value;

        public readonly ResultError? Error;

        public bool IsSuccess => Error is null;

        private Result( T value )
        {
            Value = value;
        }

        private Result( ResultError? error )
        {
            Error = error;
        }

        public static Result<T> FromSuccess( T value )
        {
            return new Result<T>( value );
        }

        public static Result<T> FromError( ResultError? error )
        {
            return new Result<T>( error );
        }

        public static Result<T> FromError( string errorText )
        {
            return new Result<T>( new ResultError( errorText ) );
        }

        public static Result<T> FromError( Result result )
        {
            return new Result<T>( result.Error );
        }
    }
}
```

```

    public class GetRecipeByIdQueryValidator( IRecipeRepository repository ) : IAsyncValidator<GetRecipeByIdQuery>
    {
        public async Task<Result> ValidationAsync( GetRecipeByIdQuery query )
        {
            if ( query.Id <= 0 )
            {
                return Result.FromError( "Id рецепта меньше нуля" );
            }

            if ( await repository.GetByIdAsync( query.Id ) == null )
            {
                return Result.FromError( "Рецепта с этим Id не существует" );
            }

            return Result.Success;
        }
    }
```



```
        public async Task<Result<GetRecipeByIdQueryDto>> HandleAsync( GetRecipeByIdQuery query )
        {
            Result validationResult = await validator.ValidationAsync( query );
            if ( !validationResult.IsSuccess )
            {
                return Result<GetRecipeByIdQueryDto>.FromError( validationResult );
            }

            Recipe foundRecipe = await recipeRepository.GetByIdAsync( query.Id );
            if ( foundRecipe is null )
            {
                return Result<GetRecipeByIdQueryDto>.FromError( "Recipe not found" );
            }
            ... дальше думаю понятно
```

VladDotMe commented on backend/Recipes/Recipes.Application/Validation/IAsyncValidator.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public interface IAsyncValidator<TData> where TData : class
+    {
+        Task<ValidationResult> ValidationAsync( TData inputData );
```
будет 
        Task<Result> ValidateAsync( TData inputData );

VladDotMe commented on backend/Recipes/Recipes.Application/Validation/IAsyncValidator.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Recipes.Application.Validation
+{
+    public interface IAsyncValidator<TData> where TData : class
```
in TData
https://learn.microsoft.com/en-us/dotnet/standard/generics/covariance-and-contravariance

Если пока сложно для понимания - можешь не вникать, просто сделай.

VladDotMe commented on backend/Recipes/Recipes.Application/Recipes/Queries/GetRecipeById/GetRecipeByIdQueryValidator.cs:
```diff
@@ -0,0 +1,30 @@
+                return ValidationResult.Fail( "Id рецепта меньше нуля" );
+            }
+
+            if ( await _repository.GetByIdAsync( query.Id ) == null )
```
Согласно nullable reference types - это выражение всегда ложь. Почитай про null ref types еще раз..

tanyginvv replied:
Нужно обсудить, опытным путем, если поставить != то найденный рецепт с Id выведет, что рецепт не найден ( соответственно он найден). Метод возвращает нам класс-ссылочный тип. Если метод не нашел нужный нам класс, то он вернет null.
![image](https://github.com/user-attachments/assets/f04a6b2b-0e64-442f-8c4a-897efff741cb)

tanyginvv replied:
Прочитал статью по **is** и **==** .
https://medium.com/@erickgallani/the-difference-between-null-check-with-is-instead-of-in-c-ed3d955a639e .
Понял, что лучше использовать оператор **is**

VladDotMe replied:
Тут вопрос был не про то, что возвращает метод репозитория (а он действительно может вернуть null), а про nullable reference types :)

tanyginvv committed:
Правки с учетом замечаний.

Поправил код с учетом замечаний.
Реализовал класс для работы с изображениями( их загрузкой и выгрузкой из store)

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Entities/Recipes/RecipeRepository.cs:
```diff
@@ -0,0 +1,58 @@
+                .Include( r => r.Ingredients )
+                .Include( r => r.Tags )
+                .Where( r => r.Id == id )
+                .SingleOrDefaultAsync();
```
FirstOrDefault

VladDotMe commented:
Договорились выключить везде nullable reference types

VladDotMe commented on backend/Recipes/Recipes.Application/Recipes.Application.csproj:
```diff
@@ -0,0 +1,17 @@
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
```
disable сделать тут и везде во всех проектах

tanyginvv committed:
выставил nullable reference type disable, добавил класс и апи для загрузки и получения картинки

tanyginvv committed:
добавление апи для поиска рецепта по тегам и названиям

tanyginvv committed:
написал апи для получения 5 случайных тегов, чтобы предлагает их в окне поиска

tanyginvv committed:
если нажимать пуск с пустой строкой, будет возвращать все рецепты

tanyginvv committed:
вынес хранилище фото в инфрастркутуру, imageTools из юзкейсов

VladDotMe commented on backend/Recipes/Recipes.Domain/Entities/Ingredient.cs:
```diff
@@ -3,24 +3,16 @@
-        public string Title { get; private set; }
-        public string Description { get; private set; }
+        public int RecipeId { get; set; }
+        public string Title { get; set; }
```
так давай все поля у которых сеттер не используется сделаем get; init; В соседних Entities аналогично.

VladDotMe commented on backend/Recipes/Recipes.Domain/Entities/User.cs:
```diff
@@ -3,35 +3,15 @@
     public class User
```
класс не используется, давай пока удалим

VladDotMe commented on backend/Recipes/Recipes.Domain/Entities/Recipe.cs:
```diff
@@ -3,44 +3,26 @@
-        public string Name { get; private set; }
-        public string Description { get; private set; }
-        public int CookTime { get; private set; }
-        public int CountPortion { get; private set; }
```
PortionCount (это я к названию прикопался) )

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,137 @@
+    [Route( "api/recipes" )]
+    public class RecipesController( IMapper mapper ) : ControllerBase
+    {
+        private IMapper _mapper => mapper;
```
Не используется, либо везде убрать маппер либо везде использовать

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,29 @@
+    public class TagsController( IQueryHandler<IReadOnlyList<Tag>, GetRandomTagsQuery> getRandomTagsQueryHandler )
+        : ControllerBase
+    {
+        private IQueryHandler<IReadOnlyList<Tag>, GetRandomTagsQuery> _getRandomTagsQueryHandler => getRandomTagsQueryHandler;
```
Ты видимо неправильно понял суть Primary Constructors)
Идея в том, чтобы уменьшить количество кода, путем неявного создания readonly field's для входных параметров, а тут получается тоже самое, что если бы был обычный конструктор. 

   Должно быть вот так:
   
```
    [ApiController]
    [Route( "api/tags" )]
    public class TagsController( IQueryHandler<IReadOnlyList<Tag>, GetRandomTagsQuery> getRandomTagsQueryHandler )
        : ControllerBase
    {
        [HttpGet]
        public async Task<IActionResult> GetDistinctTags( [FromQuery] int count = 5 )
        {
            var query = new GetRandomTagsQuery { Count = count };
            var result = await getRandomTagsQueryHandler.HandleAsync( query );

            if ( !result.IsSuccess )
            {
                return NotFound( result.Error );
            }

            return Ok( result.Value );
        }
    }
    ```

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,29 @@
+        [HttpGet]
+        public async Task<IActionResult> GetDistinctTags( [FromQuery] int count = 5 )
+        {
+            var query = new GetRandomTagsQuery { Count = count };
```
Давай использовать везде конкретные типы, тебе студия с .editorconfig не подсказывает?

Еще в .NET удобно, оно само догадается какой тип используется при создании экземпляра объекта, то есть просто вот так будет работать:
            GetRandomTagsQuery query = new() { Count = count };
            
Ну и ниже тип тоже лучше указать.
            Result<IReadOnlyList<Tag>> result = await getRandomTagsQueryHandler.HandleAsync( query );

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,29 @@
+        private IQueryHandler<IReadOnlyList<Tag>, GetRandomTagsQuery> _getRandomTagsQueryHandler => getRandomTagsQueryHandler;
+
+        [HttpGet]
+        public async Task<IActionResult> GetDistinctTags( [FromQuery] int count = 5 )
```
Но в целом непонятно:
1.Что такое DistinctTags, почему такое название?
2.Что такое GetRandomTagsQuery и почему Random? 
3. И почему названия у метода контроллера и handler'a совершенно разные?

Я пока не до конца понимаю где на UI будет использоваться этот метод, но если речь идет про этот блок
![image](https://github.com/user-attachments/assets/db8224dd-49d2-4291-ac65-4247a0bae596)

То лучше назвать типа GetFeaturedTags

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/ImagesController.cs:
```diff
@@ -0,0 +1,48 @@
+        [HttpPost( "upload" )]
+        public async Task<IActionResult> UploadImage( IFormFile image )
+        {
+            if ( image == null )
```
is null предпочтительнее чем == null
Аналогично с != null -> is not null

Разница вот в чем:
https://stackoverflow.com/a/50811687

Если вкратце: == можно переопределить, is нельзя.

VladDotMe commented on backend/Recipes/Recipes.Application/Validation/IAsyncValidator.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IAsyncValidator<in TData> where TData : class
+    {
+        Task<Result> ValidationAsync( TData inputData );
```
ValidateAsync, опечатка в названии

VladDotMe commented on backend/Recipes/Recipes.Application/Results/Result.cs:
```diff
@@ -0,0 +1,72 @@
+{
+    public record ResultError( string Message );
+
+    public class Result
```
Вынести классы по разным файлам. Еще убрать везде "?" т.к. в четверг решили отключить nullable ref types в проектах.

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Queries/GetAllRecipes/GetAllRecipesQueryHandler.cs:
```diff
@@ -0,0 +1,86 @@
+
+        public async Task<Result<IEnumerable<RecipeDto>>> HandleAsync( GetAllRecipesQuery query )
+        {
+            Result validationResult = await _recipeQueryValidator.ValidationAsync( query );
```
Заметил, что у тебя у всех Queries/Commands одинаковое начало логики в HandleAsync (везде начинается с валидации)? 
Есть идеи как это можно унифицировать?) Предлагаю тебе подумать, что можно с этим сделать и потом на одной из встреч обсудим, пока в коде не меняй.

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,137 @@
+        [HttpGet]
+        public async Task<IActionResult> GetAllRecipes( [FromServices] IQueryHandler<IEnumerable<RecipeDto>, GetAllRecipesQuery> getAllRecipesQueryHandler )
+        {
+            var result = await getAllRecipesQueryHandler.HandleAsync( new GetAllRecipesQuery() );
```
Не хватает паджинации, помнишь обсуждали, что она должна быть на бекенде)
Попробуй реализовать, подсказка: для фильтрации страниц стоит использовать паттерн "Specification". К сожалению, сходу не смог найти хорошей статьи которую можно посоветовать, предлагаю тебе самому почитать про него в интернете, тебе будет достаточно минимальной версии паттерна (В итоге нужно добавить какой-нибудь generic PaginationFilter, который можно применять к любым объектам и прокидывать его в GetAllAsync у RecipeRepository)

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Queries/GetAllRecipes/GetAllRecipesQueryHandler.cs:
```diff
@@ -0,0 +1,86 @@
+            foreach ( var recipe in recipes )
+            {
+                var stepsQuery = new GetStepsByRecipeIdQuery { RecipeId = recipe.Id };
+                var stepsResult = await _stepsQueryHandler.HandleAsync( stepsQuery );
```
Зачем вся эта логика?)) У тебя уже в recipes есть материализованные Steps/Ingredients и Tags (т.к. ты написал Include), просто используй их

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,137 @@
+        private IMapper _mapper => mapper;
+
+        [HttpPost]
+        public async Task<IActionResult> CreateRecipe( [FromBody] RecipeCreateDto dto,
```
IActionResult нельзя использовать, используй конкретный возвращаемый тип.

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,137 @@
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> GetAllRecipes( [FromServices] IQueryHandler<IEnumerable<RecipeDto>, GetAllRecipesQuery> getAllRecipesQueryHandler )
```
О а еще смотри какой момент, ты возвращаешь полные сущности рецепта когда запрашиваешь все рецепты, зачем? По сути гоняешь много лишних данных по сетке) Отдавай только то, что нужно, а то что не нужно - не отдавай) У тебя вот выше есть получение RecipeById, там отдашь полную инфу по рецепту, а здесь тебе нужно только то, что отображается в списке рецептов.

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,89 @@
+            Result validationResult = await _createRecipeCommandValidator.ValidationAsync( createRecipeCommand );
+            if ( !validationResult.IsSuccess )
+            {
+                return Result.FromError( validationResult.Error.Message );
```
Тут же можешь просто вернуть validationResult

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,89 @@
+            }
+
+            Recipe recipe = new Recipe(
+            createRecipeCommand.Name,
```
неправильное форматирование (отступы)

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Entities/Tags/TagRepository.cs:
```diff
@@ -0,0 +1,61 @@
+            _context = context;
+        }
+
+        public async Task AddAsync( Tag tag )
```
У тебя есть AddAsync в BaseRepository, тебе не нужно его заново писать здесь, аналогично в соседних репозиториях.

Еще пересмотри методы BaseRepository, они все должны использоваться. 
У тебя щас все не используются. 

В этих классах у тебя вообще не должно быть _context'a по идее, можешь ссылаться на тот который есть в BaseRepository.

VladDotMe replied:
Кстати, студия так умеет?
![image](https://github.com/user-attachments/assets/f042aab0-bb86-4a1b-8197-b55c3b9590bb)

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,89 @@
+            createRecipeCommand.ImageUrl );
+
+            await _recipeRepository.AddAsync( recipe );
+            await _unitOfWork.CommitAsync();
```
Если мы вызовем CommitAsync, но между вызовом этого CommitAsync и следующего CommitAsync ниже, случится что-нибудь непредвиденное (например: отключится электричество и комп с приложением отрубится), что будет в базе данных?) 

Это прям очень важный комментарий, подумай серьезно про всю логику вызовов здесь и ниже. Если не поймешь в чем проблема - обсудим на следующей встрече.

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,137 @@
+
+        [HttpGet( "search" )]
+        public async Task<IActionResult> SearchRecipes( [FromQuery] List<string> searchTerms,
+           [FromServices] IQueryHandler<IEnumerable<RecipeDto>, SearchRecipesQuery> searchRecipesQueryHandler )
```
Давай объединим методы GetAllRecipes и SearchRecipes а под капотом для поиска тоже будем использовать спецификацию.

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/DeleteRecipe/DeleteRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,64 @@
+                return Result.FromError( "Recipe not found" );
+            }
+
+            foreach ( var ingredientDto in foundRecipe.Ingredients.ToList() )
```
Здесь тебе не нужно вручную удалять ингредиенты/шаги. Нужно сделать так, чтобы СУБД сама это делала.
https://learn.microsoft.com/en-us/ef/core/saving/cascade-delete

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/UpdateRecipe/UpdateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,138 @@
+            var oldSteps = oldRecipe.Steps.ToList();
+            var newSteps = updateRecipeCommand.Steps.ToList();
+
+            for ( int i = 0; i < Math.Min( oldSteps.Count, newSteps.Count ); i++ )
```
Очень сложная логика нахождения дельты Steps и Ingredients, попробуй упростить. Если не получится, потом посмотрим вместе.

Nafanyan replied:
++++

tanyginvv committed:
внес правки, добавил пагинацию, со спецификацией

поправил первичные конструкторы, добавил пагинацию, внес правки, убрал лишние методы и тд

tanyginvv committed:
мини фикс

tanyginvv committed:
подправил навзание

Nafanyan commented on backend/Recipes/Recipes.Application/ImageTools/ImageHelper.cs:
```diff
@@ -0,0 +1,62 @@
+    public class ImageHelperTools
+    {
+        private const string STORE_URL = "../Recipes.Infrastructure/store";
+        public async Task<string> SaveRecipeImageAsync( IFormFile image )
```
пустая строка

Nafanyan commented on backend/Recipes/Recipes.Application/ImageTools/ImageHelper.cs:
```diff
@@ -0,0 +1,62 @@
+                return null;
+            }
+
+            var currentDirectory = Directory.GetCurrentDirectory();
```
var'ы лучше использовать только когда из правой части понятен тип переменной

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Commands/CreateIngredient/CreateIngredientCommandHandler.cs:
```diff
@@ -0,0 +1,32 @@
+                createIngredientCommand.RecipeId );
+
+            await ingredientRepository.AddAsync( ingredient );
+
```
предлагаю вызовы unitOfWork'а выносить в useCase'ы

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,79 @@
+                await createStepCommandHandler.HandleAsync( createStepCommand );
+            }
+
+            await unitOfWork.CommitAsync();
```
где-то ты вызываешь CommitAsync в useCase'ах, а где-то в реализации репозитория. Давай сделаем как здесь, CommitAsync будет вызываться в конце useCase'ов

Nafanyan commented on backend/Recipes/Recipes.Infrastructure/Entities/Ingredients/IngredientRepository.cs:
```diff
@@ -0,0 +1,45 @@
+
+        public async Task DeleteAsync( int id )
+        {
+            var ingredient = await GetByIdAsync( id );
```
var

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Commands/UpdateIngredient/UpdateIngredientCommandHandler.cs:
```diff
@@ -0,0 +1,34 @@
+                }
+                else
+                {
+                    return Result.FromError( "Такого id не существует" );
```
Какого id, id чего? Нужно более информативно

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Dtos/IngredientDtoUseCases.cs:
название какое-то странное, как минимум DTO должно быть в конце

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Queries/GetIngredientsByRecipeIdQuery/GetIngredientsByRecipeIdQueryHandler.cs:
```diff
@@ -0,0 +1,36 @@
+            IEnumerable<Ingredient> ingredients = await ingredientRepository.GetByRecipeIdAsync( query.RecipeId );
+            if ( ingredients is null )
+            {
+                return Result<GetIngredientsByRecipeIdQueryDto>.FromError( "Ingredients not found" );
```
где-то пишешь ошибки на русском, где-то на английском, давай везде на русском

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandValidator.cs:
```diff
@@ -0,0 +1,50 @@
+                return Result.FromError( "Описание блюда не может быть больше чем 150 символов" );
+            }
+
+            if ( command.PortionCount == 0 || command.PortionCount < 0 )
```
почему бы просто не использовать <= ?
```c#
if ( command.PortionCount <= 0 ) {
    return Result.FromError( "Количество порций должно быть больше 0" );
}
```

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandValidator.cs:
```diff
@@ -0,0 +1,50 @@
+                return Result.FromError( "Время приготовления должно быть больше 0" );
+            }
+
+            if ( command.ImageUrl is null || command.ImageUrl == string.Empty )
```
string.IsNullOrEmpty( command.ImageUrl ); ?

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/UpdateRecipe/UpdateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,126 @@
+            Recipe oldRecipe = await recipeRepository.GetByIdAsync( updateRecipeCommand.Id );
+            if ( oldRecipe is null )
+            {
+                return Result.FromError( "Recipe not found" );
```
Написать на русском

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Dtos/GetAllRecipesQueryDto.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace Recipes.Application.UseCases.Recipes.Dtos
+{
+    public class GetAllRecipesQueryDto
```
а оно нужно?

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Tags/Commands/CreateTag/CreateTagCommand/CreateTagCommand.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Recipes.Application.UseCases.Tags.Commands.CreateTag.CreateTagCommand
```
namespace Recipes.Application.UseCases.Tags.Commands.CreateTag.CreateTagCommand => namespace Recipes.Application.UseCases.Tags.Commands.CreateTag?

Nafanyan replied:
и как будто директория CreateTagCommand лишняя

Nafanyan commented on backend/Recipes/Recipes.Infrastructure/Entities/Ingredients/IngredientRepository.cs:
```diff
@@ -0,0 +1,45 @@
+            await base.AddAsync( ingredient );
+        }
+
+        public async Task UpdateAsync( Ingredient ingredient )
```
зачем тебе асинхронный update если update не асинхронный?

Nafanyan commented on backend/Recipes/Recipes.Infrastructure/Entities/BaseRepository.cs:
```diff
@@ -0,0 +1,35 @@
+        await _context.SaveChangesAsync();
+    }
+
+    public virtual async Task Update( TEntity entity )
```
этот метод уберем, Update будет происходить через SaveChangesAsync у unitOfWork'а

Nafanyan replied:
т.е. ты будешь получать объект сущности, менять его проперти, а потом просто вызывать SaveChangesAsync(), EFCore умный и поймет, что сущность подверглась модификации, а потому при вызове SaveChangesAsync отправит изменения в БД

Nafanyan reviewed (changes requested):
Это только начало, завтра допроверю)

Nafanyan commented on backend/Recipes/Recipes.Domain/Entities/User.cs:
откуда эти изменения? @VladTL может сделать rebase с мастером и форс пушнуть изменения?

Nafanyan commented on backend/Recipes/Recipes.Domain/Entities/Tag.cs:
```diff
@@ -3,17 +3,12 @@
-        public int RecipeId { get; init; }
-        public string Name { get; private set; }
+        public string Name { get; init; }
+        public ICollection<Recipe> Recipes { get; set; }
```
Почему в других местах используешь List<T>, а здесь ICollection<T>?

VladDotMe replied:
> Почему в других местах используешь List, а здесь ICollection?

Обычно лучше использовать наиболее абстрактный тип, в случае EFки по моему ICollection, но надо везде к одному привести, да..

Nafanyan commented on backend/Recipes/Recipes.Application/CQRSInterfaces/ICommandHandler.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Recipes.Application.CQRSInterfaces
+{
+
```
пустая строка

Nafanyan commented on backend/Recipes/Recipes.Application/ImageTools/ImageHelper.cs:
```diff
@@ -0,0 +1,62 @@
+﻿using Microsoft.AspNetCore.Http;
```
@VladTL может я чего-то не понимаю, но это логика веба, application точно должен оперировать такими либами?

VladDotMe replied:
да, здесь вообще от IFormFile нужен только FileName, можно сразу его строкой прокидывать

Nafanyan commented on backend/Recipes/Recipes.Application/ImageTools/ImageHelper.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace Recipes.Application.ImageTools
+{
+    public class ImageHelperTools
```
@VladTL по ощущениям эта штука должна быть в инфраструктуре, если поменяется файловое хранилище мы будем менять Application, т.е. бизнес логику, это не хорошо. Например если мы захотим перейти на облачное хранилище S3 то это будет смена внешней зависимости, т.е. менять надо инфраструктуру.

Предлагаю механизм сохранения и получения картинки вынести в отдельный интерфейс, реализовать в инфраструктуре, а здесь ссылаться на него

VladDotMe replied:
Согласен с @Nafanyan

Nafanyan commented on backend/Recipes/Recipes.Application/Repositories/IIngredientRepository.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Recipes.Application.Repositories
+{
+    public interface IIngredientRepository
```
@VladTL  как думаешь, может схожую логику по типу добавления, удаления вынести в отдельный интерфейс, а здесь оставить какой-то специфичный набор конкретно для этого репозитория, например вынести в 
 IAddedRepository<TEntity>
 IRemovableRepository<TEntity>
 ISearchRepository<TEntity>
 
 тогда здесь останутся только методы
 Task<IReadOnlyList<Ingredient>> GetByRecipeIdAsync( int recipeId )
 Task<Ingredient> GetByIdAsync( int id );
 
 По моему это будет соответствие одному из принципов SOLID (Принцип разделения интерфейса)

VladDotMe replied:
+++

Nafanyan commented on backend/Recipes/Recipes.Application/Repositories/IUnitOfWork.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Recipes.Application.Repositories
+{
+    public interface IUnitOfWork
```
этого не должно быть в директории Repositories, здесь только интерфейсы репизиториев

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Commands/DeleteIngredient/DeleteIngredientCommandHandler.cs:
```diff
@@ -0,0 +1,32 @@
+            var ingredient = await ingredientRepository.GetByIdAsync( command.Id );
+            if ( ingredient is null )
+            {
+                return Result.FromError( "Ingredient not found" );
```
Привести язык ошибок к единому стилю (русскому)

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Commands/DeleteIngredient/DeleteIngredientCommandValidator.cs:
```diff
@@ -0,0 +1,25 @@
+        {
+            if ( command.Id <= 0 )
+            {
+                return Result.FromError( "Invalid ingredient ID." );
```
Привести язык ошибок к единому стилю (русскому)

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Commands/UpdateIngredient/UpdateIngredientCommandHandler.cs:
```diff
@@ -0,0 +1,34 @@
+        public async Task<Result> HandleAsync( UpdateIngredientCommand updateIngredientCommand )
+        {
+            Result validationResult = await validator.ValidateAsync( updateIngredientCommand );
+            if ( !validationResult.IsSuccess )
```
если валидация была неудачна, тогда мы обновляем Ингридиент?

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Dtos/IngredientDtoUseCases.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace Recipes.Application.UseCases.Ingredients.Dtos
+{
+    public class IngredientDtoUseCases
```
IngredientDto?

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,79 @@
+                    Description = ingredientDto.Description,
+                    RecipeId = recipe.Id
+                };
+                await createIngredientCommandHandler.HandleAsync( createIngredientCommand );
```
Во первых, из-за того что ты реализовал saveChange на уровне инфраструктуры, то при 100 ингриедиентах ты 100 раз будешь сохранять изменения в БД, что не круто
Во вторых, вы же не зря реализовали проперти Ingredients и Steps в Recipe, почему бы просто не заполнять их, а потом одним  SaveChangesAsync сохранять изменения, возможно придется подправить конфиги настроки EFCore.

Короче сделать так же как ты сделал с тегами выше, что думаешь @VladTL

VladDotMe replied:
Мы обсуждали с Вадимом в личке уже, UnitOfWork в проекте используется неправильно, надо переделать)

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/DeleteRecipe/DeleteRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,36 @@
+            Recipe foundRecipe = await recipeRepository.GetByIdAsync( deleteRecipeCommand.RecipeId );
+            if ( foundRecipe is null )
+            {
+                return Result.FromError( "Recipe not found" );
```
можно я больше не буду помечать такие места, сам найдешь тексты ошибок на английском и исправишь на русский?)

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/UpdateRecipeTags/UpdateRecipeTagsCommandHandler.cs:
```diff
@@ -0,0 +1,82 @@
+
+namespace Recipes.Application.UseCases.Recipes.Commands.UpdateRecipeTags
+{
+    public class UpdateRecipeTagsCommandHandler(
```
может это стоит перенести в папку Tags?

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Dtos/GetRecipeQueryDto.cs:
```diff
@@ -0,0 +1,20 @@
+        public required string ImageUrl { get; init; }
+        public required List<TagDtoUseCases> Tags { get; init; }
+        public required List<StepDtoUseCases> Steps { get; init; }
+        public required List<IngredientDtoUseCases> Ingredients { get; init; }
```
почему бы здесь не использовать определенные в этой же директории IngredientDto, а IngredientDtoUseCases снести? Можешь объяснить зачем так сделал?

tanyginvv replied:
У меня было не помню по каким уже папкам, но одинаковые dto по названиям, и EF ругался на неоднозначность, поэтому решил сделать такой костыль, но уже правлю.

Nafanyan commented on backend/Recipes/Recipes.WebApi/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,28 @@
+using Recipes.Application.CQRSInterfaces;
+using Recipes.Application.Results;
+using Recipes.Application.UseCases.Tags.Queries.GetRandomTags;
+using Recipes.Domain.Entities;
```
у тебя по моему нет доступа к этому проекту

Nafanyan reviewed (changes requested):
ПР не рабочий, вроде как, надеюсь проблема не на моём железе. Надеюсь дальше ПРы будут поменьше)

VladDotMe commented:
@Nafanyan годные комменты, все по делу, ответил на пару моментов. 
Пока не стал смотреть все заново, @tanyginvv поправь все комменты, поправь UnitOfWork и еще раз гляну.

tanyginvv committed:
внес часть правок, но пока поправил не до конца

Nafanyan commented:
поздравляю тебя с первой соткой комментов на ПРе)

tanyginvv committed:
поправил до конца

Настроил unitOfWork.
Поправил создание и обновление рецепта, обновление массивов вынес в отдельные команды, поменял реализацию, раскидал по папкам нужным, подправил вывод ошибок и тд

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,104 @@
+        }
+
+        [HttpGet]
+        public async Task<ActionResult<IReadOnlyList<GetRecipePartDto>>> SearchRecipes(
```
Давай более универсальное название какое-нибудь? GetRecipes?

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -0,0 +1,104 @@
+
+        [HttpGet]
+        public async Task<ActionResult<IReadOnlyList<GetRecipePartDto>>> SearchRecipes(
+             [FromQuery] List<string> searchTerms,
```
А почему searchItems это коллекция а не строка? Если что, это вопрос) Может есть какой-то замысел?) А так лучше в конец аргументов и по умолчанию = null
Ну и в целом [FromServices] лучше в самый конец параметров помещать и не смешивать их с [FromQuery], сложно читается

Предлагаю придерживаться такого порядка:
1.Route data ([FromRoute])
2.Query string ([FromQuery])
3.Form data ([FromForm])
4.Request body ([FromBody])
5.HTTP headers ([FromHeader])
6.Services ([FromServices])

tanyginvv replied:
![image](https://github.com/user-attachments/assets/571352b9-6370-48be-8bf3-7fb9cff88fd8)
Пишет ошибка CS1737: Необязательные параметры должны быть указаны после всех требуемых параметров.

serchTerms это коллекция строк, потому что мы можем закидывать теги и названия несвязные друг с другом и будем получать все рецепты( использовать несколько поисковых терминов), в которых встретился тот или иной тег или название.

tanyginvv replied:
Требует такой порядок 
![image](https://github.com/user-attachments/assets/96aab3b3-3b44-4e57-b6cc-4d0950aa3eb0)

VladDotMe commented on backend/Recipes/Recipes.WebApi/Dto/RecipeDtos/RecipeCreateDto.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        [Required]
+        [StringLength( 50 )]
+        public string Name { get; set; }
```
setеры не нужны, можно везде get; init;

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Entities/Recipes/RecipeRepository.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+        }
+
+        public override async Task AddAsync( Recipe recipe )
```
зачем оверрайд?

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,95 @@
+            IUnitOfWork unitOfWork )
+        : ICommandHandlerWithResult<CreateRecipeCommand, int>
+    {
+        public async Task<Result<int>> HandleAsync( CreateRecipeCommand createRecipeCommand )
```
Нужно возвращать конкретный тип (объект), непонятно что за int..

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Tags/Commands/CreateTag/CreateTagCommandHandler.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Recipes.Application.UseCases.Tags.Commands.CreateTag
+{
+    public class CreateTagCommandHandler(
```
Я бы назвал это GetOrCreateTag..

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Tags/Commands/CreateTag/CreateTagCommandHandler.cs:
```diff
@@ -0,0 +1,36 @@
+                return Result<Tag>.FromError( validationResult.Error );
+            }
+
+            bool tagExists = await tagRepository.ContainsAsync( tag => tag.Name == createTagCommand.Name );
```
Лучше из репозитория достать первый такой объект (через FirstOrDefault)  и сразу его возвращать если он найден, чтобы дважды не ходить в базу. 

И так в целом код можно упростить, сделать вида:
Tag tag = await repo.GetByNameAsync(..);
if ( tag is null ) 
{ 
.. create tag 
}

return tag;

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,95 @@
+
+            foreach ( TagDto tagDto in createRecipeCommand.Tags )
+            {
+                CreateTagCommand createTagCommand = new() { Name = tagDto.Name };
```
Тут бы тоже маппером маппить) 

Все еще рекомендую Mapster, еще пока не сложно перейти) там проще две сущности в одну маппить (тебе это пригодится ниже)

VladDotMe replied:
В любом случае, создавать объекты прям тут посреди логики - не очень хорошая затея. Либо автомапперы либо Extension методы для перегона.

tanyginvv replied:
Пока не везде успел прикрутить Mapster, но вопрос. Получается так, что не везде получится его использовать.
Например, при создании рецепта.
![image](https://github.com/user-attachments/assets/8beab145-8db0-44e3-ae5b-262ebe61e066)
Тут мы без проблем используем Mapster.
Но в ингредиентах и шагах, мы уже так сделать не можем, тк нужно передавать рецепт в dto, но мы так сделать не можем, мы его берем после создания в этом методе.
![image](https://github.com/user-attachments/assets/7ccc3b99-3a50-404a-b393-8e68678ca497)
Маппер получилось прикрутить при получении рецептов, ну и в некоторых случаях. Пройдусь еще по проекту, может где-то еще можно дописать.

tanyginvv replied:
Таким образом, вытекает вопрос, везде ли нужно его использовать

VladDotMe commented on backend/Recipes/Recipes.WebApi/Dto/RecipeDtos/RecipeCreateDto.cs:
```diff
@@ -0,0 +1,34 @@
+        public string ImageUrl { get; set; }
+
+        [Required]
+        public ICollection<IngredientDto> Ingredients { get; set; }
```
IngredientDto и другие сущности должны быть на уровне WebApi

tanyginvv replied:
Как поступить? Если я вынесу в Api с такими же названиями dto, то они будут конфликтовать с dto из useCasов, если я уберу ссылку на юзкейсы, то в сервисах будут недоступны нужные dto и query

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,95 @@
+                {
+                    Title = ingredientDto.Title,
+                    Description = ingredientDto.Description,
+                    RecipeId = recipe.Id
```
Уверен, что здесь будет RecipeId?) Мне кажется, что нет. Как вариант - можешь сюда прокинуть сам Recipe а не его айдишник, по идее должно сработать но надо проверить)

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Entities/Steps/StepConfiguration.cs:
```diff
@@ -0,0 +1,32 @@
+            builder.HasOne( s => s.Recipe )
+                .WithMany( r => r.Steps )
+                .HasForeignKey( s => s.RecipeId )
+                .OnDelete( DeleteBehavior.Cascade );
```
А почему это правило тут а не в Recipe? Ведь мы задаем правила удаления для рецепта а не для шага.. Аналогично с Ingredient

tanyginvv replied:
Подправил, решил посмотреть в миграциях, по-моему, там ничего не изменилось, но это не точно

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Queries/GetRecipeById/GetRecipeByIdQueryHandler.cs:
```diff
@@ -0,0 +1,58 @@
+
+            Recipe foundRecipe = await recipeRepository.GetByIdAsync( query.Id );
+
+            GetRecipeByIdQueryDto getRecipeByIdQueryDto = new GetRecipeByIdQueryDto
```
Тут прям железобетон нужен маппер

VladDotMe commented on backend/Recipes/Recipes.Application/Paginator/PaginationFilter.cs:
```diff
@@ -0,0 +1,14 @@
+    public class PaginationFilter
+    {
+        public int PageNumber { get; set; } = 1;
+        public int PageSize { get; set; } = 4;
```
Property initializer value ignored during initialization

VladDotMe commented on backend/Recipes/Recipes.Application/Paginator/PaginationFilter.cs:
```diff
@@ -0,0 +1,14 @@
+﻿namespace Recipes.Application.Paginator
+{
+    public class PaginationFilter
```
Папку лучше назвать Filters и поместить рядом еще SearchFilter

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Entities/Recipes/RecipeRepository.cs:
```diff
@@ -0,0 +1,51 @@
+            }
+        }
+
+        public async Task<IReadOnlyList<Recipe>> GetAllAsync( PaginationFilter paginationFilter )
```
Это работает немного не так. Во-первых, у тебя будет всего один метод а не два. Во-вторых, он должен принимать либо список из фильтров либо заранее заданный набор фильтров (PaginationFilter .. = null, SearchFilter ...= null)

VladDotMe replied:
Вот пример от чатжпт

Step 1: Define a Common Filter Interface
Define an interface that all filters will implement.

```
public interface IFilter<T>
{
    IQueryable<T> Apply(IQueryable<T> query);
}
```
Step 2: Implement Specific Filters
Implement the SearchFilter and PaginationFilter classes to conform to the IFilter<T> interface.

```
public class SearchFilter : IFilter<Recipe>
{
    public string SearchTerm { get; set; }

    public IQueryable<Recipe> Apply(IQueryable<Recipe> query)
    {
        if (!string.IsNullOrEmpty(SearchTerm))
        {
            query = query.Where(r => r.Name.Contains(SearchTerm) ||
                                     r.Description.Contains(SearchTerm));
        }
        return query;
    }
}

```
```
public class PaginationFilter : IFilter<Recipe>
{
    public int PageNumber { get; set; } = 1;
    public int PageSize { get; set; } = 10;

    public IQueryable<Recipe> Apply(IQueryable<Recipe> query)
    {
        var skip = (PageNumber - 1) * PageSize;
        query = query.Skip(skip).Take(PageSize);
        return query;
    }
}

```
Step 3: Create an Extension Method to Apply Filters
Create an extension method that applies a collection of filters to an IQueryable

```
using System.Collections.Generic;
using System.Linq;

public static class QueryableExtensions
{
    public static IQueryable<T> ApplyFilters<T>(this IQueryable<T> query, IEnumerable<IFilter<T>> filters)
    {
        foreach (var filter in filters)
        {
            query = filter.Apply(query);
        }
        return query;
    }
}

```
Step 4: Update the Service Method
Update the service method to accept an IEnumerable<IFilter<Recipe>> and apply these filters.


```
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;

public class RecipeService
{
    private readonly YourDbContext _context;

    public RecipeService(YourDbContext context)
    {
        _context = context;
    }

    public async Task<List<Recipe>> GetRecipesAsync(IEnumerable<IFilter<Recipe>> filters)
    {
        var query = _context.Recipes.AsQueryable().ApplyFilters(filters);
        return await query.ToListAsync();
    }
}

```
Step 5: Usage Example
Here's how you can use the updated GetRecipesAsync method with multiple filters:


```
public async Task ExampleUsage()
{
    var filters = new List<IFilter<Recipe>>
    {
        new SearchFilter { SearchTerm = "chocolate" },
        new PaginationFilter { PageNumber = 1, PageSize = 10 }
    };

    var recipeService = new RecipeService(new YourDbContext());
    var recipes = await recipeService.GetRecipesAsync(filters);

    // Do something with the recipes
}

```
Explanation
Common Interface: The IFilter<T> interface defines a method Apply that takes an IQueryable<T> and returns an IQueryable<T>.
Specific Filters: The SearchFilter and PaginationFilter classes implement the IFilter<Recipe> interface, providing specific filtering logic.
Extension Method: The ApplyFilters extension method iterates over the collection of filters and applies each one to the query.
Service Method: The GetRecipesAsync method in the RecipeService class now accepts an IEnumerable<IFilter<Recipe>> and applies these filters using the extension method.
Usage: The usage example shows how to create a list of filters and pass them to the GetRecipesAsync method.
This approach makes the filtering logic highly flexible and extensible, allowing you to easily add new types of filters in the future.

VladDotMe replied:
А да, PaginationFilter можно сделать generic'ом

VladDotMe commented on backend/Recipes/Recipes.Domain/Entities/Recipe.cs:
```diff
@@ -3,44 +3,26 @@
     public class Recipe
     {
         public int Id { get; init; }
```
Самое главное чуть не забыл.. Давай вынесем базовый класс Entity и от него будут наследоваться все остальные Entities, у него будет только Id

VladDotMe reviewed (changes requested):
Я устал смотреть, остальное досмотрю потом. В целом можешь поправить то что написал, там много общих вещей которые применяются ко всему остальному тоже

tanyginvv committed:
поправил часть замечаний

tanyginvv committed:
убрал varы

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Commands/CreateIngredient/CreateIngredientCommandHandler.cs:
```diff
@@ -0,0 +1,32 @@
+                createIngredientCommand.Description,
+                createIngredientCommand.Recipe.Id );
+
+            await ingredientRepository.AddAsync( ingredient );
```
У нас подразумевается что ингридиент отдельно от рецепта создать нельзя, верно? поэтому здесь unitOfWork не вызываешь?

Если это так, то этот класс будет использоваться только в Application слое, потому как нет смысла вызывать его извне, ведь результат не сохранится из-за отсутствия вызова коммита. Сделаем класс internal? @VladTL

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Commands/UpdateIngredient/UpdateIngredientCommandHandler.cs:
```diff
@@ -0,0 +1,33 @@
+        public async Task<Result> HandleAsync( UpdateIngredientCommand updateIngredientCommand )
+        {
+            Result validationResult = await validator.ValidateAsync( updateIngredientCommand );
+            if ( validationResult.IsSuccess )
```
а если validationResult.IsSuccess = false, то в это условие ты не попадешь и выйдешь через return Result.FromSuccess(), но почему, валидацию же не прошел

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Queries/GetIngredientsByRecipeIdQuery/GetIngredientsByRecipeIdQueryHandler.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Recipes.Application.UseCases.Ingredients.Queries.GetIngredientsByRecipeIdQuery
+{
+    public class GetIngredientsByRecipeIdQueryHandler( IIngredientRepository ingredientRepository,
```
```c#
    public class GetIngredientsByRecipeIdQueryHandler( 
        IIngredientRepository ingredientRepository,
        IAsyncValidator<GetIngredientsByRecipeIdQuery> validator ) 
        : IQueryHandler<GetIngredientsByRecipeIdQueryDto, GetIngredientsByRecipeIdQuery>
```

Nafanyan commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/IngredientMappingConfig.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public class IngredientMappingConfig
+    {
+        public static void RegisterMappings()
```
что это? какой-то задел на будущее?

Nafanyan commented on backend/Recipes/Recipes.Infrastructure/store/0ae2b5df-23b4-4dc5-bde3-b2d315798543.png:
Название папки Store c большой буквы?

Nafanyan commented on backend/Recipes/Recipes.WebApi/Dto/IngredientDtos/IngredientDto.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace Recipes.WebApi.Dto.IngredientDtos
+{
+    //public class IngredientDto
```
нужно?

Nafanyan commented on backend/Recipes/Recipes.WebApi/Dto/TagDtos/TagDto.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace Recipes.WebApi.Dto.TagDtos
+{
+    //public class TagDto
```
нужно?

tanyginvv committed:
Внес правки, убрал лишнее.

Добавил логику взаимодействия с хранилищем картинок, их удалении, при badRequest создания и обновления, также удаления фотографии старой при корректном обновлении рецепта и удалении фото при удалении самого рецепта

tanyginvv committed:
поправил удаление картинки. Явно удаляем картинку, только после commitа

tanyginvv committed:
лишний var

VladDotMe commented on backend/Recipes/Recipes.WebApi/Program.cs:
```diff
@@ -0,0 +1,26 @@
+
+var builder = WebApplication.CreateBuilder( args );
+
+builder.Services.AddAutoMapper( typeof( Program ).Assembly );
```
используется? ты же вроде на Mapster  перешел?

VladDotMe commented on backend/Recipes/Recipes.WebApi/Recipes.http:
```diff
@@ -1,6 +1,6 @@
-@Recipes_HostAddress = http://localhost:5218
-
-GET {{Recipes_HostAddress}}/weatherforecast/
```
убрать

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Tags/Commands/GetOrCreateTag/GetOrCreateTagCommand.cs:
```diff
@@ -0,0 +1,9 @@
+﻿using Recipes.Domain.Entities;
```
неймспейс не тот, пройдись по проекту утилитой (в студии есть) чтобы везде неймспейсы поправила автоматом

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Tags/Commands/GetOrCreateTag/GetOrCreateTagCommandHandler.cs:
```diff
@@ -0,0 +1,36 @@
+                return Result<Tag>.FromError( validationResult.Error );
+            }
+
+            Tag existingTag = await tagRepository.GetByNameAsync( createTagCommand.Name );
```
Tag tag = await tagRepository.GetByNameAsync( createTagCommand.Name );
            if ( tag is null )
            {
                tag = createTagCommand.Adapt<Tag>();
                await tagRepository.AddAsync( tag );
            }

            return Result<Tag>.FromSuccess( tag );
            
            можно упростить)

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Context/RecipesDbContext.cs:
```diff
@@ -0,0 +1,40 @@
+            modelBuilder.ApplyConfiguration( new TagConfiguration() );
+            modelBuilder.ApplyConfiguration( new StepConfiguration() );
+
+            modelBuilder.Entity<Recipe>()
```
А ключ составной?)

VladDotMe commented on backend/Recipes/Recipes.WebApi/appsettings.json:
```diff
@@ -0,0 +1,12 @@
+  },
+  "AllowedHosts": "*",
+  "ConnectionStrings": {
+    "Recipes": "Server=(localDB)\\MSSQLLocalDB;Database=Recipes;Trusted_Connection=true;TrustServerCertificate=True;"
```
давай добавим несколько appsettings и будем запускать через launchSettings.json,

В реальных проектах у тебя будет несколько environment'ов (сред), обычно это:

dev - общий дефолтный для всех с базовыми параметрами
username - конкретно заточенный под твой юзернейм
prod - продовские данные.

Будет выглядеть примерно так:
appsettings.json
appsettings.tanygin.json
appsettings.prod.json

Таким образом, например, я смогу создать себе файл appsettings.strelnikov.json и указать там другой ConnectionString для базы данных  

А в файле launchSettings.json обычно бывает типа:
```
{
	"profiles": {
		"tanygin": {
			"commandName": "Project",
			"environmentVariables": {
				"jsonConfigName": "dev"
			}
		},
		"prod": {
			"commandName": "Project",
			"environmentVariables": {
				"jsonConfigName": "prod"
			}
		}
	}
}
```

Добавляется примерно так
```
            IConfiguration configuration = new ConfigurationBuilder()
                .AddJsonFile( "appsettings.json" )
                .AddJsonFile( $"appsettings.{Environment.GetEnvironmentVariable("jsonConfigName")}.json" )
                .Build();
                ```

VladDotMe commented:
Почисти еще лишние using's везде где можно

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Ingredients/Commands/UpdateIngredients/UpdateIngredientsCommandHandler.cs:
```diff
@@ -0,0 +1,59 @@
+
+            List<Ingredient> oldIngredients = command.Recipe.Ingredients.ToList();
+
+            foreach ( IngredientDto newIngredient in command.NewIngredients )
```
Теперь вы понимаете почему надо было Steps/Ingredients делать как ValueObject на уровне EfCore а не создавать лишние таблицы в базе @tanyginvv  @Nafanyan ?) Тут бы вообще не было всего этого, просто бы написали Recipe.Steps = newSteps;

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ImageTools/ImageHelper.cs:
```diff
@@ -0,0 +1,91 @@
+{
+    public class ImageHelperTools : IImageTools
+    {
+        private const string STORE_URL = "../Recipes.Infrastructure/store";
```
StoreUrl

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Tags/Commands/UpdateRecipeTags/UpdateTagsCommandHandler.cs:
```diff
@@ -0,0 +1,73 @@
+
+            foreach ( string name in newTagNames )
+            {
+                bool tagExists = await tagRepository.ContainsAsync( tag => tag.Name == name );
```
Зачем вначале искать потом получать? Сразу получи и проверь на is null, и ISearchableRepository тебе не нужен будет вообще. По аналогии как предлагал где-то в соседнем месте, предлагаю получить, если он null -> создаешь и дальше общий код добавления.

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Tags/Commands/UpdateRecipeTags/UpdateTagsCommandHandler.cs:
```diff
@@ -0,0 +1,73 @@
+            }
+
+            List<Tag> existingTags = recipe.Tags.ToList();
+            List<string> existingTagNames = existingTags.Select( t => t.Name ).ToList();
```
не используется?

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,95 @@
+
+            foreach ( IngredientDto ingredientDto in createRecipeCommand.Ingredients )
+            {
+                CreateIngredientCommand createIngredientCommand = new()
```
CreateIngredientCommand createIngredientCommand = ( ingredientDto, recipe ).Adapt<CreateIngredientCommand>();

попробуй так

VladDotMe replied:
и в конфиг маппингов
            TypeAdapterConfig<(IngredientDto ingredientDto, Recipe recipe), CreateIngredientCommand>.NewConfig()
                .Map( dest => dest.Title, src => src.ingredientDto.Title )
                .Map( dest => dest.Description, src => src.ingredientDto.Description )
                .Map( dest => dest.Recipe, src => src.recipe );

tanyginvv replied:
![image](https://github.com/user-attachments/assets/11af6b62-d2b7-475e-9b6d-528563546863)
Просит конструкторы по умолчанию.

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -0,0 +1,95 @@
+                createRecipeCommand.PortionCount,
+                createRecipeCommand.ImageUrl );
+
+            await recipeRepository.AddAsync( recipe );
```
можно AddASync в конец перед unitOfWork.Commit перенести

VladDotMe replied:
а отсюда не убрал

tanyginvv committed:
правки по коментам

tanyginvv committed:
мини-правки

VladDotMe commented:
Обсудили в тг, надо поправить:
1. .editorconfig. Убрать корневой и сделать на уровне бекенда для .NET и потом на уровне фронтенда для react'a и js
2. Конфиги не работают, нужно добавить в launchsettings.Json который в Properties нужные профили. IIS предлагаю выпилить и оставить только нужное (профиль для юзернейма и prod)

tanyginvv committed:
правка конфигов

tanyginvv committed:
убрал лишнее

tanyginvv committed:
Обновил .editorconfig

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ImageTools/ImageHelper.cs:
```diff
@@ -0,0 +1,91 @@
+{
+    public class ImageHelperTools : IImageTools
+    {
+        private const string StoreUrl = "../Recipes.Infrastructure/store";
```
store -> storage
store это магазин, storage - хранилище

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ImageTools/ImageHelper.cs:
```diff
@@ -0,0 +1,91 @@
+    {
+        private const string StoreUrl = "../Recipes.Infrastructure/store";
+
+        public async Task<string> SaveRecipeImageAsync( IFormFile image )
```
Давай назовем просто SaveImageAsync подразумевая, что это может быть изображение не только для рецепта

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Entities/Recipes/RecipeRepository.cs:
```diff
@@ -0,0 +1,48 @@
+{
+    public class RecipeRepository : BaseRepository<Recipe>, IRecipeRepository
+    {
+        public RecipeRepository( RecipesDbContext context ) : base( context )
```
convert to primary constructor

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Entities/Recipes/RecipeRepository.cs:
```diff
@@ -0,0 +1,48 @@
+        {
+        }
+
+        public async Task AddAsync( Recipe recipe )
```
удалить

VladDotMe commented on backend/Recipes/Recipes.Application/Repositories/BasicRepositories/IAddedRepository.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Recipes.Application.Repositories.BasicRepositories
+{
+    public interface IAddedRepository<TEntity> where TEntity : class
```
Давай назовем IAddEntityRepository

VladDotMe commented on backend/Recipes/Recipes.Application/Repositories/BasicRepositories/IRemovableRepository.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Recipes.Application.Repositories.BasicRepositories
+{
+    public interface IRemovableRepository<TEntity> where TEntity : class
```
IDeleteEntityRepository

Не нравится название IRemovableRepository т.к. это подразумевает скорее то, что сам репозиторий "удаляемый" а не имеет метод удаления сущности.

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Entities/Tags/TagRepository.cs:
```diff
@@ -0,0 +1,38 @@
+                .ToListAsync();
+        }
+
+        public async Task<IReadOnlyList<Tag>> GetTagsForSearchAsync( int count )
```
Твой метод неэффективный, ты достаешь из базы ВСЕ теги и материализуешь их. Если их там 100 миллионов - это очень тяжелая операция.  И потом ты опять проходишься по всем тегам и сортируешь их в рандомном порядке. Это очень невыгодно с точки зрения ресурсов. 

Лучше сделать так:
```
public async Task<IReadOnlyList<Tag>> GetTagsForSearchAsync(int count)
{
    return randomTags = await _dbSet
        .OrderBy(_ => Guid.NewGuid())
        .Take(count)
        .ToListAsync();
}
```

Это сконвертируется в примерно такой запрос
SELECT TOP (@__count_0) [t].[Id], [t].[Name], [t].[OtherColumns]
FROM [Tags] AS [t]
ORDER BY NEWID()

Таким образом ты дашь возможность БД сортировать в нужном порядке и получишь нужное количество элементов, СУБД сама уже все сооптимизирует как надо.

tanyginvv committed:
правки по комментам( форматирование, лишние репозитории и методы)

VladDotMe reviewed (approved):
Сливай)

tanyginvv committed:
сделал новые неймспейсы
